[squeezebox] Remove notification volume channel and replace with config property

mhilbush opened the pull request:
Signed-off-by: Mark Hilbush <mark@hilbush.com>

Removes `notificationSoundVolume` channel and replaces it with a configuration parameter on the Player thing. This maintains consistency with pending changes to Sonos, as well as supports the modifications to audio actions where the volume can be specified as part of the audio and voice APIs.

Also adds a `notificationTimeout` config parameter on the Player thing to provide greater user control over notification duration, as well as to maintain consistency with Sonos notification functionality.

Depends on next stable ESH containing this PR .
https://github.com/eclipse/smarthome/pull/4963

Also
- Rework HttpUtils to not throw Exception in order to resolve build failure due to code analysis error
- Rework Notification Player to consolidate code for player state restoration (see https://github.com/openhab/openhab2-addons/pull/3231#discussion_r168277844)
- Fixed play/pause channel not updating when mode changes
- Fix player not pausing after notification completes
- Replace Runnables with lambdas

mhilbush commented:
Ugh, This is not what I wanted to do. I was trying to rebase my changes in this PR on the current master in order to incorporate the squeezebox bugfix to properly resume player state. @martinvw I am not sure how to fix this.

mhilbush commented:
@martinvw This is ready for review.

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxNotificationPlayer.java:
```diff
@@ -104,7 +90,7 @@ private void setupPlayerForNotification() throws InterruptedException, SqueezeBo
         int notificationVolume = squeezeBoxPlayerHandler.getNotificationSoundVolume().intValue();
         if (notificationVolume == 0) {
-            logger.warn("Player notification volume is 0.");
+            logger.info("Player notification volume is 0.");
```
Yes. We had done it originally to give the user a clue why they might not be hearing the notification. I think that's less of an issue now that the separate notification volume channel is gone.

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/README.md:
```diff
@@ -99,27 +98,48 @@ The Selection widget in HABpanel can be used to present the favorites as a choic
+
+You can set the default notification volume in the player thing configuration.
+
+You can override the default notification volume by supplying it as a parameter to *say* and *playSound*
```
Good catch. That does appear to be the convention.

mhilbush commented:
Thanks for the review @cweitkamp!

mhilbush commented:
@martinvw  With eclipse/smarthome#4963 merged and in the latest ESH stable, this is ready for review.

cweitkamp commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxPlayerHandler.java:
```diff
@@ -592,12 +591,9 @@ int currentRepeat() {
-                if (playing) {
-                    updateChannel(mac, CHANNEL_CURRENT_PLAYING_TIME, new DecimalType(currentTime++));
-                }
+        Runnable runnable = () -> {
```
You can write this in a single line instead of defining a variable. 
```
timeCounterJob = scheduler.scheduleWithFixedDelay(() -> {
    if (playing) {
        updateChannel(mac, CHANNEL_CURRENT_PLAYING_TIME, new DecimalType(currentTime++));
    }
}, 0, 1, TimeUnit.SECONDS);
```

cweitkamp commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/internal/discovery/SqueezeBoxPlayerDiscoveryParticipant.java:
```diff
@@ -110,11 +110,8 @@ private boolean playerThingExists(ThingUID newThingUID) {
-            public void run() {
-                squeezeBoxServerHandler.requestPlayers();
-            }
+        Runnable runnable = () -> {
```
Same here.
```
requestPlayerJob = scheduler.scheduleWithFixedDelay(() -> {
    squeezeBoxServerHandler.requestPlayers();
}, 10, TTL, TimeUnit.SECONDS);
```

cweitkamp commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/handler/SqueezeBoxServerHandler.java:
```diff
@@ -111,13 +111,7 @@ public SqueezeBoxServerHandler(Bridge bridge) {
-                connect();
-            }
-        }, 0, TimeUnit.SECONDS);
+        scheduler.schedule(this::connect, 0, TimeUnit.SECONDS);
```
You should use `scheduler.submit(...)` in this case.

mhilbush commented:
Thanks @cweitkamp

mhilbush commented:
@wborn Would appreciate if you can get this in your queue for a review. Thanks!

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/ESH-INF/thing/thing-types.xml:
```diff
@@ -95,6 +94,15 @@
+			<parameter name="notificationTimeout" type="integer" unit="s" required="true">
+				<label>Notification Timeout</label>
+				<description>Maximum amount of time in seconds for which the notification will be played</description>
+				<default>90</default>
```
Agreed. I will lower it.

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/ESH-INF/thing/thing-types.xml:
```diff
@@ -95,6 +94,15 @@
 				<label>MAC Address</label>
 				<description>SqueezeBox Players are identified by their MAC address</description>
 			</parameter>
+			<parameter name="notificationTimeout" type="integer" unit="s" required="true">
```
Probably not. As long as there is a default specified (which there is in this case), required is not needed. Correct?

wborn replied:
That's correct. It would also allow for omitting the parameter in .things files and adding a comment about it being required in the README. :-)

mhilbush replied:
Good point about the README. I'll update that, too.

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/internal/config/SqueezeBoxPlayerConfig.java:
```diff
@@ -12,8 +12,22 @@
+    /**
+     * Number of seconds to wait to time out a notification
+     */
+    public Integer notificationTimeout;
```
Yes, I can do that here.

mhilbush commented on addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/internal/config/SqueezeBoxPlayerConfig.java:
```diff
@@ -12,8 +12,22 @@
+    /**
+     * Volume used for playing notifications
+     */
+    public Integer notificationVolume;
```
I understand, however, given that this parameter is not required AND there is no default, if I make it an `int`, how do I tell the difference between there being no value versus a value of 0? In other words, if it's defined as an `int`, what will getConfigAs(SqueezeBoxPlayerConfig.class).notificationVolume return if the config parameter has no value?

wborn replied:
Yes you're right! Let's just leave it an Integer so it is possible to see it has no value and the `currentVolume()` can be used.

mhilbush commented:
@wborn Thanks for the feedback. I think I've addressed all your comments.

I fixed a formatting issue in thing-types.xml, and it is now showing a conflict. I'm not sure how to resolve that.

wborn commented:
> I fixed a formatting issue in thing-types.xml, and it is now showing a conflict. I'm not sure how to resolve that.

Spaces in all XML files have been converted to tabs on the latest master (see https://github.com/openhab/openhab2-addons/pull/3311). So putting the spaces back should easily resolve the merge conflict. Another way could be to rebase the branch on the current master.

mhilbush commented:
I put the spaces back. I'll fix that in the next PR. ;-)

Jenkins is not happy, but it looks like the ZWay binding is causing the heartburn?

wborn commented:
Thanks @mhilbush! We need to get https://github.com/openhab/openhab2-addons/issues/3402 fixed to resolve these compilation issues in the ZWay binding.
